docs/dev/windows.md

```diff
@@ -622,133 +690,138 @@ input
 </div>
 </div>
 
-## Dealing with Late Data
+## Triggers
 
-When working with event-time windowing it can happen that elements arrive late, i.e the
-watermark that Flink uses to keep track of the progress of event-time is already past the
-end timestamp of a window to which an element belongs. Please
-see [event time](./event_time.html) and especially
-[late elements](./event_time.html#late-elements) for a more thorough discussion of
-how Flink deals with event time.
+A `Trigger` determines when a window (as formed by the `WindowAssigner`) is ready to be
+processed by the *window function*. Each `WindowAssigner` comes with a default `Trigger`. 
+If the default trigger does not fit your needs, you can specify a custom trigger using `trigger(...)`.
 
-You can specify how a windowed transformation should deal with late elements and how much lateness
-is allowed. The parameter for this is called *allowed lateness*. This specifies by how much time
-elements can be late. Elements that arrive within the allowed lateness are still put into windows
-and are considered when computing window results. If elements arrive after the allowed lateness they
-will be dropped. Flink will also make sure that any state held by the windowing operation is garbage
-collected once the watermark passes the end of a window plus the allowed lateness.
+The trigger interface provides five methods that react to different events: 
 
-<span class="label label-info">Default</span> By default, the allowed lateness is set to
-`0`. That is, elements that arrive behind the watermark will be dropped.
+* The `onElement()` method is called for each element that is added to a window. 
+* The `onEventTime()` method is called when  a registered event-time timer fires. 
+* The `onProcessingTime()` method is called when a registered processing-time timer fires. 
+* The `onMerge()` method is relevant for stateful triggers and merges the states of two triggers when their corresponding windows merge, *e.g.* when using session windows. 
+* Finally the `clear()` method performs any action needed upon removal of the corresponding window. 
 
-You can specify an allowed lateness like this:
+Any of these methods can be used to register processing- or event-time timers for future actions. 
 
-<div class="codetabs" markdown="1">
-<div data-lang="java" markdown="1">
-{% highlight java %}
-DataStream<T> input = ...;
+### Fire and Purge
 
-input
-    .keyBy(<key selector>)
-    .window(<window assigner>)
-    .allowedLateness(<time>)
-    .<windowed transformation>(<window function>);
-{% endhighlight %}
-</div>
+Once a trigger determines that a window is ready for processing, it fires. This is the signal for the window operator to emit the result of the current window. Given a window with a `WindowFunction` 
```

Comment:
Maybe we should mention earlier which trigger methods can return a `TriggerResult` and quickly mention what that means. "A trigger fires when it returns FIRE or FIRE_AND_PURGE."